File: Rational.java
/** Defining a Rational number class
Refer to the instructions on Canvas for more information.
"I have neither given nor received help on this assignment."
author: Josh Collins
*/

public class Rational {
	// These are the instances variables defined for your class.  Don't change these lines!
	private int numerator;
	private int denominator;
	
	/** 
	 * Initialize a new Rational object with value iNum/iDen stored in the numerator and
	 * denominator variables.  Calls the reduce() method to put the fraction in lowest terms.
	 */

//	Constructor that initializes the numerator and denominator
	public Rational(int iNum, int iDen) {
		this.numerator = iNum;
		this.denominator = iDen;		
		this.reduce();
	}

//	Getter for numerator
	public int getNumerator() {
		return this.numerator;
	}

//	Setter for numerator
	public void setNumerator(int n) {
		this.numerator = n;
		this.reduce();
	}

//	Getter for denominator
	public int getDenominator() {
		return this.denominator;
	}

//	Setter for denominator
	public void setDenominator(int d) {
		this.denominator = d;
		this.reduce();
	}

//	Reciprocal method, returns the flip of a rational fraction
	public void reciprocal() {
//		Double-checking the denominator can't be 0
		if (this.numerator == 0) {
			throw new Error("Error, can't have denominator as 0");
		}
//		Creates new var for opposite values, returns reduced form
		int newDen = this.denominator;
		int newNum = this.numerator;
		this.numerator = newDen;
		this.denominator = newNum;
		this.reduce();
	}

//	Checks validity of a fraction
	public boolean isValid() {
		// Instead of always returning true, this method should return false under the right conditions
		if (this.getDenominator() != 0) {
			return true;
		} else {
			return false;
		}
	}

//	Add method, returns rational fraction plus new num
	public void add(Rational num2) {
//		Creates new val for numerator and denominator using basic fraction math
		int newDen = this.getDenominator() * num2.getDenominator();
		int newNum = this.getNumerator() * num2.getDenominator() + this.getDenominator() * num2.getNumerator();
		this.denominator = newDen;
		this.numerator = newNum;
//		Reduces to gcf
		this.reduce();
	}

//	Subtract method, returns rational fraction minus new num
	public void sub(Rational num2) {
//		Creates new val for numerator and denominator using basic fraction math
		int newDen = this.getDenominator() * num2.getDenominator();
		int newNum = this.getNumerator() * num2.getDenominator() - this.getDenominator() * num2.getNumerator();
		this.denominator = newDen;
		this.numerator = newNum;
//		Reduces to gcf
		this.reduce();
	}

//	Multiply method, returns rational fraction times new num
	public void mult(Rational num2) {
//		Creates new val for numerator and denominator using basic fraction math
		int newDen = this.getDenominator() * num2.getDenominator();
		int newNum = this.getNumerator() * num2.getNumerator();
		this.denominator = newDen;
		this.numerator = newNum;
//		Reduces to gcf
		this.reduce();
	}

//	Divide method, returns rational fraction over new num
	public void div(Rational num2) {
//		Checks that numerator can't be 0, or else number DNE
		if (num2.getNumerator() == 0) {
			throw new Error("Can't divide by 0");
		}
//		Creates new val for numerator and denominator using basic fraction math
		int newDen = this.getDenominator() * num2.getNumerator();
		int newNum = this.getNumerator() * num2.getDenominator();
		this.denominator = newDen;
		this.numerator = newNum;
//		Reduces to gcf
		this.reduce();
	}
	
	/*******************************
	*    HELPER FUNCTIONS BELOW    *
	*******************************/
	
	/**
	  Reduces the Rational to lowest terms
        - Checks if both the numerator and denominator are negative; if so, makes both positive
        - Calls gcf() to find the greatest common factor between the numerator and denominator, and
          continues to divide by that gcf until the greatest common factor is 1
	*/
	private void reduce() {
		if ((this.numerator < 0) && (this.denominator < 0)) {
			this.numerator = -this.numerator;
			this.denominator = -this.denominator;
		} //if
		
		int common = 0;
		while (common != 1) {
			common = this.gcf();
			this.numerator /= common;
			this.denominator /= common;
		} //while
		
	} //reduce
	
	/**
      Determines the greatest common factor between the numerator and denominator
        - Starts checking numbers counting downward from the smaller of the numerator,denominator pair
        - When it finds a number divisble by both, it breaks the loop and returns that number
        - The smallest number that can be returned is 1
	*/	
	private int gcf() {
		int commonFactor = 1;
		for (int i = Math.min(Math.abs(this.numerator), Math.abs(this.denominator)); i > 1; i--) {
			if ((this.numerator % i == 0) && (this.denominator % i == 0)) {
				commonFactor = i;
				break;
			} //if
		} //for
		return commonFactor;
	} //gcf
	
	/**
	  Returns a string representation of the Rational, e.g. "1/8"
	*/
	public String toString() {
		return Integer.toString(this.numerator) + "/" + Integer.toString(this.denominator);
	} //toString
	
	/**
	  Determines if two Rationals are exactly equal to each other (same numerator and same 
	  denominator, no consideration of reducing the numbers)
	*/
	public boolean equals(Object other) {
		Rational fraction = (Rational) other;
		if ((this.numerator == fraction.getNumerator()) && this.denominator == fraction.getDenominator()) {
			return true;
		} //if
		return false;
	} //equals
	
	/*******************************
	*     END HELPER FUNCTIONS     *
	*******************************/	
	
	/**
	 * This is the main function.  Running the Rational.java file will run this method. 
	 * No need for changes. It is better to put your testing in RationalTest and running that!
	 */
	public static void main(String[] args) {
		System.out.println("You're running the Rational.java file.");
		System.out.println("If you want to test your program,");
		System.out.println("you should run the RationalTest.java file instead!");
	}
} //Rational (class)

File: rational.py
""" Defining a Rational number class
Refer to the instructions on Canvas for more information.

"I have neither given nor received help on this assignment."
author: Josh Collins """

class Rational:
    """ The Rational class allows us to implement rational numbers with exact precision, 
    without the approximations/errors used in binary representations """
     
    def __init__(self, iNum, iDen):
        """ Constructs a new Rational object with value iNum/iDen stored in hidden __numerator 
        and __denominator variables. Calls reduce() to put the fraction in lowest terms. """
        self.__numerator = iNum
        self.__denominator = iDen
        self.reduce()
    
    def getNumerator(self):
        """Gets numerator from constructor"""
        return self.__numerator
    
    def getDenominator(self):
        """Gets denominator from constructor"""
        return self.__denominator
    
    def setNumerator(self, n):
        """Sets numerator as object n"""
        self.__numerator = n
        self.reduce()
            
    def setDenominator(self, d):
        """Sets denominator as object d, checks if denominator is non-zero"""
        self.__denominator = d
        self.reduce()

    def isValid(self):
        """Checks if the fraction is valid for operator functions"""
        if self.getDenominator() != 0:
            return True
        else:
            return False
    
    def reciprocal(self):
        """Returns reciprocal of fraction as long as numerator is non-zero"""
        if self.__numerator == 0:
            raise ValueError('Reciprocal of 0 is undefined')
        # Calls back to numerator and denominator and flips them
        newDen = self.__denominator
        newNum = self.__numerator
        # Assigns back to numerator and denominator
        self.__numerator = newDen
        self.__denominator = newNum
        # Returns gcf
        self.reduce()
        return self
    
    def add(self, num2):
        """Adds a number to the fraction to create a new value"""
        # Calls back to numerator and denominator and adds them with a new num
        newDen = self.getDenominator() * num2.getDenominator()
        newNum = self.getNumerator() * num2.getDenominator() + self.getDenominator() * num2.getNumerator()
        # Assigns back to numerator and denominator
        self.__denominator = newDen
        self.__numerator = newNum
        # Returns gcf
        self.reduce()
  	
    def sub(self, num2):
        """Subtracts a number from the fraction to create a new value"""
        # Calls back to numerator and denominator and subtracts from them with a new num
        newDen = self.getDenominator() * num2.getDenominator()
        newNum = self.getNumerator() * num2.getDenominator() - self.getDenominator() * num2.getNumerator()
        # Assigns back to numerator and denominator
        self.__denominator = newDen
        self.__numerator = newNum
        # Returns gcf
        self.reduce()
             
    def mult(self, num2):
        """Multiplies a number to the fraction to create a new value"""
        # Calls back to numerator and denominator and multiplies them with a new num
        newDen = self.getDenominator() * num2.getDenominator()
        newNum = self.getNumerator() * num2.getNumerator()
        # Assigns back to numerator and denominator
        self.__denominator = newDen
        self.__numerator = newNum
        # Returns gcf
        self.reduce()

    def div(self, num2):
        """Divides a number from the fraction to create a new value"""
        # Checks if the num can be divisible first
        if num2.getNumerator() == 0:
            raise ValueError('Cant divide by 0')
        # Calls back to numerator and denominator and divides them with a new num
        newDen = self.getDenominator() * num2.getNumerator()
        newNum = self.getNumerator() * num2.getDenominator()
        # Assigns back to numerator and denominator
        self.__denominator = newDen
        self.__numerator = newNum
        # Returns gcf
        self.reduce()
    
    ################################
    #    HELPER FUNCTIONS BELOW    #
    ################################

    def reduce(self):
        """ Reduces the Rational to lowest terms
        - Checks if both the numerator and denominator are negative; if so, makes both positive
        - Calls gcf() to find the greatest common factor between the numerator and denominator, and
            continues to divide by that gcf until the greatest common factor is 1 """
        if self.__numerator < 0 and self.__denominator < 0:
            self.__numerator = -self.__numerator
            self.__denominator = -self.__denominator
        common = 0
        while common != 1:
            common = self.gcf()
            self.__numerator /= common
            self.__denominator /= common
    
    def gcf(self):
        """ Determines the greatest common factor between the numerator and denominator
        - Starts checking numbers counting downward from the smaller of the numerator,denominator pair
        - When it finds a number divisble by both, it breaks the loop and returns that number
        - The smallest number that can be returned is 1 """
        common_factor = 1
        for i in range(min(abs(int(self.__numerator)), abs(int(self.__denominator))), 1, -1):
            if self.__numerator % i == 0 and self.__denominator % i == 0:
                 common_factor = i
                 break
        return common_factor
    
    def __str__(self):
        """ Returns a string representation of the Rational, e.g. "1/8" """
        return str(int(self.__numerator)) + "/" + str(int(self.__denominator))
    
    def __eq__(self, r2):
        """ Determines if two Rationals are exactly equal to each other 
        (same numerator and same denominator, no consideration of reducing the numbers) """
        return self.__numerator == r2.__numerator and self.__denominator == r2.__denominator
    
    ################################
    #     END HELPER FUNCTIONS     #
    ################################    
    
def main():
    """ This main should only start rationalTest's unit tests.  
    Do NOT implement your tests here.  WebCAT wants them in rationalTest.py """
    import unittest
    unittest.main("rationalTest")

if __name__ == "__main__":
    main()